### PR TITLE
ARROW-1324: [C++] Support ARROW_BOOST_VENDORED on Windows / MSVC

### PR DESCRIPTION
**Pull request by BACtaki**
Adding a check to use the correct executable for vendor boostcpp libraries on Windows.

JIRA tracking issue: https://issues.apache.org/jira/browse/ARROW-1324
Author: Badrul Chowdhury (badrulchowdhury17@gmail.com)
Reviewer: @wesm

### CODE REVIEW

File: cpp/cmake_modules/ThirdpartyToolchain.cmake [hunk 1]
```diff
@@ -512,15 +512,34 @@ macro(build_boost)
...
   endif()
-  set(BOOST_BUILD_COMMAND "./b2" "link=${BOOST_BUILD_LINK}"
-                          "variant=${BOOST_BUILD_VARIANT}" "cxxflags=-fPIC")
+  if(MSVC)
+    set(BOOST_BUILD_COMMAND
+        "b2.exe"
+        "link=${BOOST_BUILD_LINK}"
+        "variant=${BOOST_BUILD_VARIANT}"
+        "address-model=64"
+        "toolset=msvc-14.0"
```
Comment (wesm) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake]: Can the MSVC version need to match the current toolset in use? If we are using vs2017 or vs2019 then the toolset version number probably needs to be correspondingly higher here

Comment (bkietz) [reply cpp/cmake_modules/ThirdpartyToolchain.cmake]: CMake has a built-in variable `MSVC_TOOLSET_VERSION` which can help here
```suggestion
  if(MSVC)
    if(MSVC_TOOLSET_VERSION EQUAL 142)
      set(TOOLSET "msvc-14.2")
    elseif(MSVC_TOOLSET_VERSION EQUAL 141)
      set(TOOLSET "msvc-14.1")
    elseif(MSVC_TOOLSET_VERSION EQUAL 140)
      set(TOOLSET "msvc-14.0")
    else()
      message(SEND_ERROR "unsupported MSVC toolset version, MSVC_TOOLSET_VERSION=${MSVC_TOOLSET_VERSION}") 
    endif()
```

**Comment by bkietz** [comment]
@BACtaki it looks like CI is green except for cmake formatting; could you `run-cmake-format`?

**Comment by wesm** [comment]
Rebased and fixed formatting. Someone probably needs to try this manually since it isn't being run in CI at the moment

**Comment by wesm** [comment]
@BACtaki are you able to complete this patch?

**Comment by wesm** [comment]
Looks like @kou is taking this over

**Comment by kou** [comment]
#5337

**Pull request closed without merging**